federation/pkg/kubefed/init/init.go

```diff
@@ -206,38 +216,47 @@ func initFederation(cmdOut io.Writer, config util.AdminConfig, cmd *cobra.Comman
 	}
 
 	// 6. Create federation API server
-	_, err = createAPIServer(hostClientset, initFlags.FederationSystemNamespace, serverName, image, serverCredName, pvc.Name, advertiseAddress)
+	_, err = createAPIServer(hostClientset, initFlags.FederationSystemNamespace, serverName, image, serverCredName, pvc.Name, advertiseAddress, dryRun)
 	if err != nil {
 		return err
 	}
 
 	// 7. Create federation controller manager
-	_, err = createControllerManager(hostClientset, initFlags.FederationSystemNamespace, initFlags.Name, cmName, image, cmKubeconfigName, dnsZoneName)
+	_, err = createControllerManager(hostClientset, initFlags.FederationSystemNamespace, initFlags.Name, cmName, image, cmKubeconfigName, dnsZoneName, dryRun)
 	if err != nil {
 		return err
 	}
 
 	// 8. Write the federation API server endpoint info, credentials
 	// and context to kubeconfig
-	err = updateKubeconfig(config, initFlags.Name, endpoint, entKeyPairs)
+	err = updateKubeconfig(config, initFlags.Name, endpoint, entKeyPairs, dryRun)
 	if err != nil {
 		return err
 	}
 
-	return printSuccess(cmdOut, ips, hostnames)
+	if !dryRun {
+		return printSuccess(cmdOut, ips, hostnames)
+	} else {
+		_, err := fmt.Fprintf(cmdOut, "Federation control plane runs (dry run)\n")
+		return err
+	}
 }
 
-func createNamespace(clientset *client.Clientset, namespace string) (*api.Namespace, error) {
+func createNamespace(clientset *client.Clientset, namespace string, dryRun bool) (*api.Namespace, error) {
 	ns := &api.Namespace{
 		ObjectMeta: api.ObjectMeta{
 			Name: namespace,
 		},
 	}
 
-	return clientset.Core().Namespaces().Create(ns)
+	if !dryRun {
+		return clientset.Core().Namespaces().Create(ns)
+	} else {
```

Comment:
Go recommends dropping unnecessary `else` blocks. See https://golang.org/doc/effective_go.html#if. Drop the `else` here and just return nil, nil. Btw, I would rather write this as

```golang
if dryRun {
    return ns, nil
}

return clientset.Core().Namespace().Create(ns)
```

This reads more direct, straightforward and hence more readable.